src/qt/transactiongraphwidget.cpp

```diff
@@ -0,0 +1,376 @@
+// Copyright (c) 2019 The Bitcoin Unlimited developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "transactiongraphwidget.h"
+#include "clientmodel.h"
+#include "txmempool.h" // for TX_RATE_RESOLUTION_MILLIS
+
+#include <QColor>
+#include <QPainter>
+
+#include <cmath>
+
+/* Minutes to milliseconds conversion factor */
+static const long MINUTES_TO_MILLIS = 60 * 1000;
+/* Maximum sample window of 1 day, in milliseconds */
+static const long MAXIMUM_SAMPLE_WINDOW_MILLIS = 24 * 60 * MINUTES_TO_MILLIS;
+/* Sample rate, derived from the signal frequency used to update the TPS label on the debug ui */
+static const long SAMPLE_RATE_MILLIS = TX_RATE_RESOLUTION_MILLIS;
+/* Keep no more than this many samples in memory (older samples will be purged) */
+static const long MAXIMUM_SAMPLES_TO_KEEP = MAXIMUM_SAMPLE_WINDOW_MILLIS / SAMPLE_RATE_MILLIS;
+
+/* Always display at least 1 tps range on the y-axis */
+static const float MINIMUM_DISPLAY_YVALUE = 1.0f;
+
+/* Maximum redraw frequency */
+static const long MAXIMUM_REDRAW_RATE_MILLIS = 500;
+/* Minimum redraw frequency */
+static const long MINIMUM_REDRAW_RATE_MILLIS = 5000;
+
+static const int XMARGIN = 10;
+static const int YMARGIN = 10;
+
+TransactionGraphWidget::TransactionGraphWidget(QWidget *parent)
+    : QWidget(parent), nMinutes(1440), nRedrawRateMillis(SAMPLE_RATE_MILLIS), fDisplayMax(MINIMUM_DISPLAY_YVALUE),
+      fInstantaneousTpsPeak_Runtime(0.0f), fInstantaneousTpsPeak_Sampled(0.0f), fInstantaneousTpsPeak_Displayed(0.0f),
+      fInstantaneousTpsAverage_Runtime(0.0f), fInstantaneousTpsAverage_Sampled(0.0f),
+      fInstantaneousTpsAverage_Displayed(0.0f), fSmoothedTpsPeak_Runtime(0.0f), fSmoothedTpsPeak_Sampled(0.0f),
+      fSmoothedTpsPeak_Displayed(0.0f), fSmoothedTpsAverage_Runtime(0.0f), fSmoothedTpsAverage_Sampled(0.0f),
+      fSmoothedTpsAverage_Displayed(0.0f), nTotalSamplesRuntime(0L), vInstantaneousSamples(), vSmoothedSamples(),
+      clientModel(0)
+{
+}
+
+long TransactionGraphWidget::getSamplesInDisplayWindow() const
+{
+    return (nMinutes * MINUTES_TO_MILLIS) / SAMPLE_RATE_MILLIS;
+}
+
+void TransactionGraphWidget::setClientModel(ClientModel *model)
+{
+    clientModel = model;
+    if (model)
+    {
+        connect(model, SIGNAL(transactionsPerSecondChanged(double, double, double)), this,
+            SLOT(setTransactionsPerSecond(double, double, double)));
+    }
+}
+
+void TransactionGraphWidget::paintPath(QPainterPath &avgPath, QPainterPath &peakPath)
+{
+    int h = height() - YMARGIN * 2, w = width() - XMARGIN * 2;
+    int x = XMARGIN + w, yAvg, yPeak;
+    int samplesInWindow = getSamplesInDisplayWindow();
```

Comment:
use max(1, getSamplesInDisplayWindow()) to ensure no div by zero on line 80

Reply:
#1960 
I corrected in a different manner than using std::max() though.